xdsl/dialects/snitch_runtime.py

```diff
@@ -235,6 +235,231 @@ class DmaWaitAllOp(SnitchRuntimeBarrier):
+    Configure an SSR data mover for a 4D loop nest.
+    b0: Inner-most bound (limit of loop)
+    b3: Outer-most bound (limit of loop)
+    s0: increment size of inner-most loop
+    """
+
+    name = "snrt.ssr_loop_4d"
+    dm: Operand = operand_def(ssr_dm)
+    b0: Operand = operand_def(IndexType)
+    b1: Operand = operand_def(IndexType)
+    b2: Operand = operand_def(IndexType)
+    b3: Operand = operand_def(IndexType)
+    i0: Operand = operand_def(IndexType)
+    i1: Operand = operand_def(IndexType)
+    i2: Operand = operand_def(IndexType)
+    i3: Operand = operand_def(IndexType)
+
+    def __init__(
+        self,
+        dm: Operation | SSAValue,
+        b0: Operation | SSAValue,
+        b1: Operation | SSAValue,
+        b2: Operation | SSAValue,
+        b3: Operation | SSAValue,
+        i0: Operation | SSAValue,
+        i1: Operation | SSAValue,
+        i2: Operation | SSAValue,
+        i3: Operation | SSAValue,
+    ):
+        super().__init__(operands=[dm, b0, b1, b2, b3, i0, i1, i2, i3], result_types=[])
+
+
+@irdl_op_definition
+class SsrRepeatOp(SnitchRuntimeBaseOp, ABC):
+    """
+    Configure the repetition count for a stream.
+    """
+
+    name = "snrt.ssr_repeat"
+    dm: Operand = operand_def(ssr_dm)
+    count: Operand = operand_def(IndexType)
+
+    def __init__(
+        self,
+        dm: Operation | SSAValue,
+        count: Operation | SSAValue,
+    ):
+        super().__init__(operands=[dm, count], result_types=[])
+
+
+@irdl_op_definition
+class SsrEnableOp(SnitchRuntimeBaseOp, ABC):
+    """
+    Enable SSR.
+    """
+
+    name = "snrt.ssr_enable"
+
+    def __init__(
+        self,
+    ):
+        super().__init__(operands=[], result_types=[])
+
+
+@irdl_op_definition
+class SsrDisableOp(SnitchRuntimeBaseOp, ABC):
+    """
+    Disable SSR.
+    """
+
+    name = "snrt.ssr_disable"
+
+    def __init__(
+        self,
+    ):
+        super().__init__(operands=[], result_types=[])
+
+
+@irdl_op_definition
+class SsrReadOp(SnitchRuntimeBaseOp, ABC):
+    """
+    Start a streaming read.
+    """
+
+    name = "snrt.ssr_read"
+    dm: Operand = operand_def(ssr_dm)
+    dim: Operand = operand_def(ssr_dim)
+    ptr: Operand = operand_def(i32)
+
+    def __init__(
+        self,
+        dm: Operation | SSAValue,
+        dim: Operation | SSAValue,
+        ptr: Operation | SSAValue,
+    ):
+        super().__init__(operands=[dm, dim, ptr], result_types=[])
+
+
+@irdl_op_definition
+class SsrWriteOp(SnitchRuntimeBaseOp, ABC):
+    """
+    Start a streaming write.
+    """
+
+    name = "snrt.ssr_write"
+    dm: Operand = operand_def(ssr_dm)
+    dim: Operand = operand_def(ssr_dim)
+    ptr: Operand = operand_def(i32)
+
+    def __init__(
+        self,
+        dm: Operation | SSAValue,
+        dim: Operation | SSAValue,
+        ptr: Operation | SSAValue,
+    ):
+        super().__init__(operands=[dm, dim, ptr], result_types=[])
+
+
+@irdl_op_definition
+class FpuFenceOp(SnitchRuntimeBaseOp, ABC):
```

Comment:
Both `SSREnableOp`, `SSRDisableOp`, and `FpuFenceOp` have the exact same signature, I feel like they could inherit from the same base class (see an example with `SnitchRuntimeGetInfo`)
I think `SnitchRuntimeBarrier` implements the exact same behaviour?
Maybe we can make a more abstract `NoOperandNoResult`, or `SnitchRuntimeSideEffectOperation` base class that covers all of these classes' behaviour? That would greatly reduce the code needed to describe all of these ops

Reply:
should I also change the operations that use SnitchRuntimeBarrier then?